### PR TITLE
Slightly refactor CLI

### PR DESCRIPTION
**Pull request by whatyouhide**

### CODE REVIEW

File: lib/protobuf/protoc/cli.ex [hunk 1]
```diff
@@ -13,85 +12,106 @@ defmodule Protobuf.Protoc.CLI do
...
+          do: Protobuf.Protoc.Generator.generate(ctx, desc)
 
     response = Google.Protobuf.Compiler.CodeGeneratorResponse.new(file: files)
+
     IO.binwrite(Protobuf.Encoder.encode(response))
   end
 
+  def main(_args) do
+    fail_and_halt("Invalid arguments. See protoc-gen-elixir --help.")
+  end
```
Comment (whatyouhide) [inline lib/protobuf/protoc/cli.ex:63]: Improved UX

File: lib/protobuf/protoc/cli.ex [hunk 2]
```diff
@@ -13,85 +12,106 @@ defmodule Protobuf.Protoc.CLI do
...
+
+  # Made public for testing.
   @doc false
-  def parse_params(ctx, params_str) when is_binary(params_str) do
-    params = String.split(params_str, ",")
-    parse_params(ctx, params)
+  def parse_params(%Protobuf.Protoc.Context{} = ctx, params_str) when is_binary(params_str) do
+    params_str
+    |> String.split(",")
+    |> Enum.reduce(ctx, &parse_param/2)
```
Comment (whatyouhide) [inline lib/protobuf/protoc/cli.ex:70]: I think this reads a lot easier with as a reduce

File: lib/protobuf/protoc/cli.ex [hunk 3]
```diff
@@ -13,85 +12,106 @@ defmodule Protobuf.Protoc.CLI do
...
-    parse_params(ctx, t)
+  defp parse_param("gen_descriptors=" <> value, ctx) do
+    case value do
+      "true" ->
+        %{ctx | gen_descriptors?: true}
+
+      other ->
+        fail_and_halt(
+          "Invalid value for gen_descriptors option, expected \"true\", got: #{inspect(other)}"
+        )
```
Comment (whatyouhide) [inline lib/protobuf/protoc/cli.ex:85]: Improved UX

File: lib/protobuf/protoc/cli.ex [hunk 4]
```diff
@@ -13,85 +12,106 @@ defmodule Protobuf.Protoc.CLI do
...
   end
 
   @doc false
-  def find_types(ctx, [], acc), do: %{ctx | global_type_mapping: acc}
+  def find_types(ctx, descs) do
+    global_type_mapping =
+      Enum.reduce(descs, %{}, fn desc, acc ->
+        types = find_types_in_proto(ctx, desc)
+        Map.put(acc, desc.name, types)
+      end)
```
Comment (whatyouhide) [inline lib/protobuf/protoc/cli.ex:112]: Reduce instead of recursion here

File: lib/protobuf/protoc/cli.ex [hunk 5]
```diff
@@ -116,7 +136,7 @@ defmodule Protobuf.Protoc.CLI do
   end
 
   defp find_types_in_proto(types, ctx, %Google.Protobuf.DescriptorProto{name: name} = desc) do
-    new_ctx = append_ns(ctx, name)
+    new_ctx = Map.update!(ctx, :namespace, &(&1 ++ [name]))
```
Comment (whatyouhide) [inline lib/protobuf/protoc/cli.ex:139]: Cut out some code

File: lib/protobuf/protoc/cli.ex (hunk 4 above)
Comment (ericmj) [reply lib/protobuf/protoc/cli.ex:112]:
```suggestion
    global_type_mapping =
      Map.new(descs, fn desc ->
        {desc.name, find_types_in_proto(ctx, desc)}
      end)
```

**Review by ericmj — APPROVED** [review]

**Review by britto — COMMENTED** [review]
Neat! 💜

**Pull request merged**